### PR TITLE
fix: restrict generated m2m relationships

### PR DESCRIPTION
**Pull request by steve-chavez**
Fixes https://github.com/PostgREST/postgrest/issues/2070.

Now we only add m2m relationship based on [strict junctions](https://en.wikipedia.org/wiki/Associative_entity):

- A table has FKs to two other tables
- **These FK columns are part of the table's PK columns.**

### CODE REVIEW

**Comment by steve-chavez** [comment]
Some stats regarding the big schema(apflora):

- On the main branch, we have 179940 relationships detected - where 144396 are m2m relationships.
- On this PR, we now have 35544 relationships detected - where none are m2ms. It makes sense since no table has a composite PK.

This now reduces the `dump-schema` output to `8.6M` from the previous [67M](https://github.com/PostgREST/postgrest/pull/2260#issue-1212641135).

**Comment by steve-chavez** [comment]
> 35544 relationships

These are still a lot considering `apflora` has 45 tables and 281 views. Moving forward we probably need a way to annotate(through a `COMMENT` or other mechanism) a VIEW and only in that case generation relationships for it.

File: test/spec/fixtures/schema.sql [hunk 1]
```diff
@@ -1955,12 +1952,13 @@ create table whatev_sites (
...
 
 create table whatev_jobs (
-  job_id        uuid    primary key
+  job_id        uuid
 , name text
 , site_id_1     int not null references whatev_sites (id)
 , project_id_1  int not null references whatev_projects (id)
 , site_id_2     int not null references whatev_sites (id)
 , project_id_2  int not null references whatev_projects (id)
+, primary key(job_id, site_id_1, project_id_1, site_id_2, project_id_2)
```
Comment (steve-chavez) [inline test/spec/fixtures/schema.sql:1961]: This PK was added so this test keeps passing:

https://github.com/PostgREST/postgrest/blob/9bdb4e5c4f723c01f1a77f7b03575c7f8efb368f/test/spec/Feature/Query/EmbedDisambiguationSpec.hs#L121-L157

I do still wonder if this type of modeling can happen in the real world. Still, keeping this for completion.

File: test/spec/fixtures/schema.sql [hunk 2]
```diff
@@ -2294,7 +2292,8 @@ A test for partitioned tables$$;
         car_dealer_city varchar(64) not null,
         quantity int not null,
         foreign key (car_model_name, car_model_year) references test.car_models (name, year),
-        foreign key (car_dealer_name, car_dealer_city) references test.car_dealers (name, city)
+        foreign key (car_dealer_name, car_dealer_city) references test.car_dealers (name, city),
+        primary key (car_model_name, car_model_year, car_dealer_name, car_dealer_city, quantity)
```
Comment (steve-chavez) [inline test/spec/fixtures/schema.sql:2296]: I believe this new detection makes perfect sense. It also accounts for cases where the PK has an additional column(`quantity`) that is not part of an FK.

**Review by laurenceisla — APPROVED** [review]

**Comment by bolerodan** [comment]
Are there any builds or nightly builds where I can test these recent changes (like there was for the Ltree merge)?

**Comment by steve-chavez** [comment]
The loadtest reports a drop in performance here: https://github.com/PostgREST/postgrest/runs/6142982808. 

Rate: 445(main) vs 407(head).

Not sure why though. The change here is only related to the schema cache and it reduces the amount of relationships searched on each request.

**Pull request merged**